### ISSUE TITLE
Can't seem to work with Events/Actions

### ISSUE THREAD

**Issue by sholsapp** [StevenHeinrich/ember-leaflet-draw#4]
Hi there, I've defined following the instructions from README's "Working with Events/Actions" section, but haven't been able to get this to work. I've tried defining a custom leaflet-map componen and  a custom draw-control. Both attempts had actions defined for `onDrawCreated`, yet the event was never handled.

My goal is to be able to add a event handler that can export geojson of the created shape.

Any tips?

**Comment by StevenHeinrich** [comment]
@sholsapp I just updated things a little. I updated the events to have the event return more than just the default from the underlying Leafelt.Draw library. The first argument on the events is still the event return object (as defined by Leaflet.Draw https://leaflet.github.io/Leaflet.draw/docs/leaflet-draw-latest.html#l-draw-event-event). The second argument is the LayerGroup that contains all the drawn features. The third argument is the map object itself (b/c sometimes having total control is just what you want).

After making the update to the events... I updated the example on the home page of the dummy app. Still would be nice to fill up the examples route with some good stuff, but that will take some time and I wanted to get something back to you to get you going for now. I tried adding some explanation in the comments. Let me know if anything is vague. Documentation is certainly a work in progress and much better when read and commented on by more than just the author.

Let me know if you have a specific situation where you would like to see an example. 

The examples are on the dummy home page.

> *Template*
> https://github.com/StevenHeinrich/ember-leaflet-draw/blob/master/tests/dummy/app/templates/home.hbs
> 
> *Controller*
> https://github.com/StevenHeinrich/ember-leaflet-draw/blob/master/tests/dummy/app/controllers/home.js

**Comment by sholsapp** [comment]
I think that https://stevenheinrich.github.io/ember-leaflet-draw/ isn't loading now.

**Comment by StevenHeinrich** [comment]
Yeah, the Travis build failed... so it didn't deploy the dummy app to the GitHub page. 

~I don't have time to look into it right now. Looks like you can still clone/download the addon and build it locally to see the Dummy page.~

Nevermind... now that I blew away my local node-modules and rebuilt, I see the errors on my environment. Will try to look into it more this evening.

**Comment by sholsapp** [comment]
No rush, thanks!

**Comment by sholsapp** [comment]
P.s., I ended up find a way to work around things after digging through the code. What I did was end up writing a custom component (as follows).

```
import Ember from 'ember';                                                                                                                                                                                            
import DrawControl from 'ember-leaflet-draw/components/draw-control';                                                                                                                                                 
                                                                                                                                                                                                                      
export default DrawControl.extend({                                                                                                                                                                                   
    /**                                                                                                                                                                                                               
     * Callback provided by the parent controller or component.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             
     */                                                                                                                                                                                                               
    onLayerAdd: null,                                                                                                                                                                                                 
    /**                                                                                                                                                                                                               
     * Method called by ember-leaflet-draw's component.                                                                                                                                                                                                                                                                                                                                                     
     */                                                                                                                                                                                                               
    _drawCreated: function(e) {                                                                                                                                                                                       
        if (this.get('onLayerAdd') != null) {                                                                                                                                                                         
            this.sendAction('onLayerAdd', e);                                                                                                                                                                         
        }                                                                                                                                                                                                             
    },                                                                                                                                                                                                                
}); 
```
The big takeaway for me was that I had to override `_drawCreated`, which wasn't clear from the docs. Is what I did awful for any reason? :)

**Comment by StevenHeinrich** [comment]
I set the leaflet dependency to explicitly use 1.0.3. They just updated to 1.1.0 and thats messing some things up. There is a deprecation warning and some other things. Ember-Leaflet hasn't had a travis build since the day before this Leaflet 1.1.0 release (which was yesterday). My guess is that when Ember-Leaflet resolves things to user the latest version o Leaflet, I can piggy-back that. Their work may solve it all. If not, I can look at the changes they make to play happy with 1.1.0 and replicate within this addon.

tl:dr The builds are happy and the demo site is deployed... https://stevenheinrich.github.io/ember-leaflet-draw/

**Comment by StevenHeinrich** [comment]
In regard to extending the component... I am not exactly sure what you were looking to do, but... all of the Leaflet.Draw events are exposed as actions on the draw-control component. I think you can avoid extending the component by taking advantage of the events being exposed by actions, described a little better in the example below. I will try to get more of this kind of detail into the docs.

The handlebars example below shows how to catch the `draw:created` event from LeafLet.Draw, by binding your action to the draw-control component's action.

`onDrawCreated=(action 'updateDrawnFeature')`

In this case, when LeafletDraw fires the `draw:created` event, it will be picked up by the draw-control component and exposed on its `onDrawCreated` action. You can hook your action in there.


```handlebars
{{#leaflet-map lat=lat lng=lng zoom=zoom as |layers|}}
  {{layers.tile url="http://{s}.basemaps.cartocdn.com/light_all/{z}/{x}/{y}.png"}}
  {{draw-control position="topright"
                 draw=drawConfig
                 showDrawingLayer=true
                 enableEditing=true
                 onDrawCreated=(action 'updateDrawnFeature')}}
{{/leaflet-map}}
```


```javascript
...
  actions: {
    // All Events passed as Actions will contain two arguments after the first that is defined by Leaflet.Draw
    // 1st argument in this case is feature (as returned by [Leaflet.Draw](https://leaflet.github.io/Leaflet.draw/docs/leaflet-draw-latest.html#l-draw-event-event))
    // 2nd argument is the current LayerGroup that the draw-control component is working with
    // 3rd argument is the map object, returned because it can be useful to have access to the map object, at times
    updateDrawnFeature(feature, drawingLayerGroup, map) {
      // EXAMPLE working with 1st argument...
      // Convert the feature object to GeoJSON
      let latestFeature = feature.layer.toGeoJSON();
      this.set('latestFeature', latestFeature);

      // EXAMPLE working with 2nd argument...
      // Run Later (like timeout) to stall, due to listening to the same event that adds the feature (technially a layer) to the LayerGroup.
      // If you access the 2nd arguement `drawingLayerGroup` without waiting, its highly likely that the new feature will be missing from the LayerGroup (hasn't been added yet).
      // Ember.run.later is not needed to work with the returned `feature` right away.
      Ember.run.later((() => {
        let allFeatures = drawingLayerGroup.toGeoJSON();
        this.set('allFeatures', allFeatures);
      }), 0);

      // EXAMPLE working with 3rd argument...
      // In case you need access to the map object, it is passed in as the third argument.
      // This could be used to change the zoom based on your newly created feature, example follows.
      let zoom = map.getZoom() + 1; // Zoom in one level
      let latlng;
      if(feature.layerType === 'marker') {
        latlng = feature.layer.getLatLng();
      } else {
        latlng = feature.layer.getBounds().getCenter();
      }
      map.setView(latlng, zoom);
    }
...
```

**Comment by sholsapp** [comment]
Hah, that's _exactly_ what I wanted to do. Thanks.

**Comment by StevenHeinrich** [comment]
Awesome! Glad it helped. You can check into some of the detail behind this example (if you want more context) by viewing it on the home page of the demo/dummy app.

On thing to note on the events -> actions... is the naming convention. They all follow this pattern 'onDraw' + '{F}istletterofremainingstring'. The "D" in draw is caps and the tricky part is ONLY the first letter or the remaining event name is caps for the action name.

Example: the event "draw:editmove" sends an action named "onDrawEditmove"

**Comment by sholsapp** [comment]
The snag that I ran into after (quickly) reading the docs originally was that I didn't pass in the action to the component. Rather, I defined it on the component/controller and expected it to "just work". Passing it in is a no brainer, and makes sense now. Perhaps worth updating the documentation to have an example of you passing in a callback like this, to make it blatantly obvious.

Thanks again, this is a cool addon!
